library/src/blas2/rocblas_hpr2.hpp

```diff
@@ -0,0 +1,150 @@
+/* ************************************************************************
+ * Copyright 2016-2020 Advanced Micro Devices, Inc.
+ * ************************************************************************ */
+#pragma once
+#include "handle.h"
+#include "rocblas.h"
+
+template <typename T, typename U>
+__device__ void hpr2_kernel_calc(bool        upper,
+                                 rocblas_int n,
+                                 U           alpha,
+                                 const T*    x,
+                                 rocblas_int incx,
+                                 const T*    y,
+                                 rocblas_int incy,
+                                 T*          AP)
+{
+    rocblas_int tx = hipBlockIdx_x * hipBlockDim_x + hipThreadIdx_x;
+    rocblas_int ty = hipBlockIdx_y * hipBlockDim_y + hipThreadIdx_y;
+
+    int index = upper ? ((ty * (ty + 1)) / 2) + tx : ((ty * (2 * n - ty + 1)) / 2) + (tx - ty);
+
+    if(upper ? ty < n && tx < ty : tx < n && ty < tx)
+    {
+        AP[index] += alpha * x[tx * incx] * conj(y[ty * incy])
+                     + conj(alpha) * y[tx * incy] * conj(x[ty * incx]);
+    }
+    else if(tx == ty && tx < n)
+    {
+        AP[index] = std::real(AP[index]) + alpha * x[tx * incx] * conj(y[ty * incy])
+                    + conj(alpha) * y[tx * incy] * conj(x[ty * incx]);
+    }
+}
+
+template <rocblas_int DIM_X, rocblas_int DIM_Y, typename TScal, typename TConstPtr, typename TPtr>
+__global__ void rocblas_hpr2_kernel(bool           upper,
+                                    rocblas_int    n,
+                                    TScal          alphaa,
+                                    TConstPtr      xa,
+                                    ptrdiff_t      shift_x,
+                                    rocblas_int    incx,
+                                    rocblas_stride stride_x,
+                                    TConstPtr      ya,
+                                    ptrdiff_t      shift_y,
+                                    rocblas_int    incy,
+                                    rocblas_stride stride_y,
+                                    TPtr           APa,
+                                    ptrdiff_t      shift_A,
+                                    rocblas_stride stride_A)
+{
+    rocblas_int num_threads = hipBlockDim_x * hipBlockDim_y * hipBlockDim_z;
+    if(DIM_X * DIM_Y != num_threads)
+        return; // need to launch exactly the number of threads as template parameters indicate.
```

Comment:
Is this a check on the compiler?   Or guard against reuse?

Reply:
I don't like it. No error message, no status, nothing.

Reply:
Actually, following a discussion with Scott, we should make the kernel crashing.

Reply:
The check seems redundant to me, could reach out and see why it was initially added into gemv in the first place.